tests/unit/lms/views/api/sync_test.py

```diff
@@ -0,0 +1,99 @@
+import pytest
+
+from lms.models import HGroup
+from lms.views.api.sync import sync
+
+
+@pytest.mark.parametrize(
+    "is_learner,canvas_api_client_method_name,expected_groups",
+    [
+        pytest.param(
+            False,
+            "course_sections",
+            [
+                HGroup(
+                    name="Section 1",
+                    authority_provided_id="d0f36006728f2277f228b74c8a7f620305bfb3e7",
+                ),
+                HGroup(
+                    name="Section 2",
+                    authority_provided_id="d99674b9700f4a40a2b301d2949b61339c58236c",
+                ),
+                HGroup(
+                    name="Section 3",
+                    authority_provided_id="6a85e3651705dee4da0805b8985472343cbea94e",
+                ),
+            ],
+            id="For non-learners we expect it to create groups for all three "
+            "of the course's sections",
+        ),
+        pytest.param(
+            True,
+            "authenticated_users_sections",
+            [
+                HGroup(
+                    name="Section 2",
+                    authority_provided_id="d99674b9700f4a40a2b301d2949b61339c58236c",
+                ),
+            ],
+            id="The learner is only a member of section 2 so when the learner "
+            "launches the app we only expect it to create the group for "
+            "section 2",
+        ),
+    ],
+)
+def test_it(
+    pyramid_request,
+    canvas_api_client,
+    lti_h_service,
+    is_learner,
+    canvas_api_client_method_name,
+    expected_groups,
+):
+    if is_learner:
+        pyramid_request.lti_user = pyramid_request.lti_user._replace(roles="Learner")
+    else:
+        pyramid_request.lti_user = pyramid_request.lti_user._replace(roles="Instructor")
+
+    returned_groups = sync(pyramid_request)
+
+    canvas_api_client_method = getattr(canvas_api_client, canvas_api_client_method_name)
+    canvas_api_client_method.assert_called_once_with("test_custom_canvas_course_id")
+
+    lti_h_service.sync.assert_called_once_with(expected_groups)
+
+    authority = pyramid_request.registry.settings["h_authority"]
+    assert returned_groups == [
+        f"group:{group.authority_provided_id}@{authority}" for group in expected_groups
+    ]
```

Comment:
I think this would be better as two tests instead of a parameterised one.

This would remove:

 * The `is_learner` path which is specific to one case on either side
 * The `canvas_api_client_method getattr` stuff for a direct assertion
 * The parameterisation and param id

Is this also an opportunity for factory use to collapse the large `HGroup` declarations?

Reply:
Split this into two separate tests with an assertion helper.

I wasn't using factories here because the tests actually do pass all two kwargs to `HGroup` and actually do depend on the values of both kwargs. But a factory is still potentially beneficial because, if a new required kwarg is added to `HGroup` in future that these tests don't care about, if they're using a factory the tests won't need to be updated. So I've changed it to a factory as well

Reply:
https://github.com/hypothesis/lms/pull/1680/commits/35dd24f0de7e482f9d613a7f240cf09871d6bca3